File: atcoder.jp/PyPy3 (7.3.0)/abc286/abc286_e.py
from collections import defaultdict, deque

def bfs(graph, n, a):
    depth = defaultdict(lambda : [-1, -1])
    depth[n][0] = 0
    depth[n][1] = a[n]
    queue = deque()
    queue.append(n)

    while queue:
        u = queue.popleft()
        for v in graph[u]:
            if depth[v][0] == -1:
                queue.append(v)
                depth[v][0] = depth[u][0] + 1
                depth[v][1] = depth[u][1] + a[v]
            elif depth[v][0] == depth[u][0] + 1:
                depth[v][1] = max(depth[v][1], depth[u][1] + a[v])

    return depth

N = int(input())
A = list(map(int, input().split()))
graph = defaultdict(set)

for i in range(N):
    S = list(input())
    for j in range(N):
        if S[j] == "Y":
            graph[i].add(j)

Q = int(input())
for _ in range(Q):
    U, V = map(int, input().split())
    U -= 1
    V -= 1

    depth = bfs(graph, U, A)
    d, v = depth[V]
    if d == -1:
        print("Impossible")
    else:
        print(d, v)

File: atcoder.jp/Rust (1.42.0)/abc286/abc286_e.rs
use proconio::{
    fastout, input,
    marker::{Chars, Usize1},
};
use std::collections::{HashMap, HashSet};

fn bfs(
    graph: &std::collections::HashMap<usize, HashSet<usize>>,
    start: usize,
    a: &[i64],
) -> std::collections::HashMap<usize, (usize, i64)> {
    let mut queue = std::collections::VecDeque::new();
    let mut visited = std::collections::HashMap::new();
    queue.push_back(start);
    visited.insert(start, (0, a[start]));
    while let Some(node) = queue.pop_front() {
        let cost = visited[&node];
        for &next in graph.get(&node).unwrap_or(&HashSet::new()).iter() {
            if !visited.contains_key(&next) {
                visited.insert(next, (cost.0 + 1, cost.1 + a[next]));
                queue.push_back(next);
            } else if visited[&next].0 == cost.0 + 1 {
                let current = visited[&next].1;
                visited.insert(next, (cost.0 + 1, std::cmp::max(current, cost.1 + a[next])));
            }
        }
    }
    visited
}

#[fastout]
fn main() {
    input! {
        n: usize,
        a: [i64; n],
    }

    let mut graph = HashMap::new();
    for i in 0..n {
        input! {
            s: Chars,
        }
        for (j, c) in s.iter().enumerate() {
            if *c == 'Y' {
                graph.entry(i).or_insert(HashSet::new()).insert(j);
            }
        }
    }

    input! {
        q: usize,
    }

    for _ in 0..q {
        input! {
            u: Usize1,
            v: Usize1,
        }

        let depth = bfs(&graph, u, &a);
        if depth.contains_key(&v) {
            let (d, v) = depth[&v];
            println!("{} {}", d, v);
        } else {
            println!("Impossible");
        }
    }
}
